src/diskquota_utility.c

```diff
@@ -1678,32 +1678,26 @@ DiskquotaShmemInitHash(const char           *name,       /* table string name fo
 		infoP->hash = oid_hash;
 	else
 		infoP->hash = string_hash;
-	return ShmemInitHash(name, init_size, max_size, infoP, hash_flags | HASH_FUNCTION);
+	return ShmemInitHash(name, init_size, max_size, infoP, hash_flags | HASH_FUNCTION | HASH_FIXED_SIZE);
 #else
-	return ShmemInitHash(name, init_size, max_size, infoP, hash_flags | HASH_BLOBS);
+	return ShmemInitHash(name, init_size, max_size, infoP, hash_flags | HASH_BLOBS | HASH_FIXED_SIZE);
 #endif /* GP_VERSION_NUM */
 }
 
 // Add or find an entry in a hash table with a size limit. If the limit is reached, only the search will be performed.
 // When overflowing, the warning warning_message will be report. But not more often than specified in
 // diskquota_hashmap_overflow_report_timeout. The time of the last warning is passed in last_overflow_report.
 void *
-shm_hash_enter(HTAB *hashp, void *keyPtr, bool *foundPtr, uint max_size, const char *warning_message,
-               time_t *last_overflow_report)
+shm_hash_enter(HTAB *hashp, const void *keyPtr, bool *foundPtr, int max_size, const char *warning_message,
+               TimestampTz *last_overflow_report)
 {
-	if (hash_get_num_entries(hashp) >= max_size)
+	void *result = hash_search(hashp, keyPtr, HASH_ENTER_NULL, foundPtr);
+	if (hash_get_num_entries(hashp) >= max_size &&
+	    TimestampDifferenceExceeds(*last_overflow_report, GetCurrentTimestamp(),
```

Comment:
try to avoid call GetCurrentTimestamp twice: save it to local var and then use it